src/object.d

```diff
@@ -2223,15 +2223,22 @@ T rehash(T : shared Value[Key], Value, Key)(T* aa)
  * Params:
  *      aa =     The associative array.
  */
-V[K] dup(T : V[K], K, V)(T aa)
+auto dup(T : V[K], K, V)(T aa)
 {
     //pragma(msg, "K = ", K, ", V = ", V);
 
     // Bug10720 - check whether V is copyable
     static assert(is(typeof({ V v = aa[K.init]; })),
         "cannot call " ~ T.stringof ~ ".dup because " ~ V.stringof ~ " is not copyable");
 
-    V[K] result;
+    // bug 11725: return mutable AA if possible
+    import core.internal.traits : Unconst;
+    static if (is(V : Unconst!V) && is(K : Unconst!K))
+        alias Result = Unconst!V[Unconst!K];
```

Comment:
This should probably be done for `K` and `V` independently.

Furthermore this should handle indirections by only stripping head const, e.g. `const int*` should be replaced with `const(int)*`. This also raises the question on how to handle structs with indirections:

```
struct S
{
    int* ptr;
}

void foo(const S[int] aa)
{
    auto copy = aa.dup();
    *(copy[0].ptr) = 2; // Can write const data
}

Reply:
Technically, keys should always be const. In fact, D makes sure of that for indirections:

```D
int[char[]] aa;
pragma(msg, typeof(aa)); // int[const(char)[]]
```

And the `is(V : Unconst!V)` should take care of the structs with indirection cases.

Reply:
You're right, Unconst was more elegant than I remembered.

Reply:
Well, `Unconst` could be dumb or poorly implemented :) It's the `is` expression that makes the difference.